### DISCUSSION TITLE
Adding NonConvex Flow objects to the transformer component

### DISCUSSION THREAD

**Discussion by eikohi** [SESMG/SESMG#176] [Q&A]
Hello everyone,

I would like to implement NonConvex Flow objects, such as minimum_downtime, for transformer components in SESMG. 
Their description for the oemof.solph version 0.4.5, which is currently used in SESMG, can be found in this [link](https://oemof-solph.readthedocs.io/en/v0.4.5/reference/oemof.solph.html#oemof.solph.options.NonConvex.).
A description of the general Flow objects, including the NonConvex object, can be found [here](https://oemof-solph.readthedocs.io/en/v0.4.5/reference/oemof.solph.html#module-oemof.solph.network.flow.).

I tried to add a line of code to the Transformer component in the preprocessing folder of SESMG, based on this [oemof.solph example](https://github.com/oemof/oemof-examples/blob/master/oemof_examples/oemof.solph/v0.4.x/min_max_runtimes/min_max_runtimes.py.)

```
    def get_primary_output_data(self, tf, max_invest=None):
        """ """
        if max_invest and max_invest < tf["max. investment capacity"]:
            tf["max. investment capacity"] = max_invest
        return {
            "outputs": {
                self.busd[tf["output"]]: Flow(
                    #new codeline
                    nonconvex=solph.NonConvex(minimum_downtime=4, initial_status=0),
                    #
                    variable_costs=tf["variable output costs"],
                    emission_factor=tf["variable output constraint costs"],
                    investment=Investment(
                        ep_costs=tf["periodical costs"],
                        minimum=tf["min. investment capacity"],
                        maximum=tf["max. investment capacity"],
                        periodical_constraint_costs=tf["periodical constraint costs"],
                        existing=tf["existing capacity"],
                        nonconvex=True if tf["non-convex investment"] == 1 else False,
                        offset=tf["fix investment costs"],
                        fix_constraint_costs=tf["fix investment constraint costs"],
                    ),
                )
            }
        }
```
However, when I try to run the optimization, I receive the following error:
 
KeyError: 'sequences'
    model.name="Model";
      - termination condition: infeasible
      - message from solver: Model was proven to be infeasible.
 
 
I'm not sure if I added the line of code in the wrong place, or if the code is fine and my model_definition is the problem. Can anyone help me? My goal is to add the minimum downtime constraint to the compression_heat_transformer component.
 
Thank you in advance
Eiko

**Comment by chrklemm** [comment]
Dear @eikohi,
thank you for contributing to the SESMG-Forum!

Generally: if you are trying to implement new features, it would be nice, if you could create a fork of the entire repository and include your changes there, so that we can see where you changed the code. Here is a tutorial how you can do that: https://docs.github.com/de/get-started/quickstart/fork-a-repo

Beside that, there are a few possible reasons why your code isn't working:

1. you included fix values, that means that your change applies for ALL transformers of your model. Try including a new column in the transformer-sheet of the model definition with any name you like ("new-column" in the following). Than you can include the value of this column in the code:
`nonconvex=solph.NonConvex(minimum_downtime=tf["new-column-1"], initial_status=tf["new-column-2"])`

2. " termination condition: infeasible" means, that the model is not solvable, either because the security of supply is not guaranted, or because there is to much energy in the system which can not be used: https://spreadsheet-energy-system-model-generator.readthedocs.io/en/latest/03.00.00_trouble_shooting.html#error-m-014-user-warning-condition-infeasible

Feel free, to keep us up to date! If this helps you or you find another way to solve this problem it would be nice if you could contributre your changes to the SESMG by creating a pull request: https://docs.github.com/de/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

Best regards, @GregorBecker & @chrklemm

**Reply by eikohi** [reply]
Hi @chrklemm and @GregorBecker,

Thank you for your quick response and your advide!
 
I forked the [repository](https://github.com/eikohi/SESMG/blob/master/program_files/preprocessing/components/Transformer.py) and changed the Codeline to make sure, that only certain transformers are manupilated.
The added code starts in line 146.
```
                    nonconvex=solph.NonConvex(
                        minimum_downtime=tf["minimum downtime"],
                        initial_status=tf["initial status"],
                    ) if "minimum downtime" in tf and not pd.isna(tf["minimum downtime"]) else None,
```
I also tried out another NonConvex Flow object:
```
                    nonconvex=solph.NonConvex(
                        startup_costs=tf["startup costs"],
                    ) if "startup costs" in tf and not pd.isna(tf["startup costs"]) else None,
```
Unfortunately, I still recieved the infeasibility error, even when I add another Investment option to fullfill the demand. Could there be a problem with the solph.NonConvex function?
 
Kind regards
Eiko

**Reply by GregorBecker** [reply]
Hi @eikohi , 

can you add your model definition file to your forked repository so that we are able to reproduce and try to fix your error. 

Thank you in advance and kind regards.

@GregorBecker

**Reply by eikohi** [reply]
Hi @GregorBecker,

I updated the code a bit (I added the NonConvex function in the imports) and uploaded my model definition sheet.
It is still not running.

Thank you and best regards
@eikohi

**Reply by GregorBecker** [reply]
Hi @eikohi , 

First of all, sorry for the late response to your request. It unfortunately got lost a bit due to many deadlines here. 

Thank you for uploading the necessary files as discussed. Since I don't have write permissions in your repository I have to describe the error handling here. If you have any questions, feel free to contact me.

I have now installed them on my system and was able to reproduce your error. I saw that you refer to the NonConvex component with solph.NonConvex. Since NonConvex is already imported at the beginning of the file, you can omit the **solph** at this point, so nonconvex=NonConvex(...) is sufficient.

After I made this change, I got an error message from oemof. This indicates that the simultaneous use of periodical costs and the nonconvex investment is not allowed. However, we do not have much experience in this area and must refer you to the oemof documentation (https://oemof-solph.readthedocs.io/en/latest/index.html). 

We hope we could help you. 

Greetings also on behalf of @chrklemm 

@GregorBecker

**Reply by eikohi** [reply]
Hi @GregorBecker,

thank you for your help! I will try it out and let you know when I figure it out.

Best regards 
@eikohi
